[IO-711] Use Objects.requireNotNull for fail fast method/constructors

boris-unckel opened the pull request:
Fixes https://issues.apache.org/jira/browse/IO-711

boris-unckel commented:
@garydgregory Please review and mark the cases you don't want to change for compatibility and/or design reasons. I'll change then all occurrences accordingly.

garydgregory commented on src/main/java/org/apache/commons/io/FileUtils.java:
```diff
@@ -866,6 +868,8 @@ public static void copyFile(final File srcFile, final File destFile, final CopyO
      */
     public static long copyFile(final File input, final OutputStream output) throws IOException {
+        Objects.requireNonNull(input, "input");
+        Objects.requireNonNull(output, "output");
```
-1: no need to check on `output`, this already done in `copyLarge`.

I'll stop this type of comment since these checks over the top in this PR. IOW, let's not check for NPEs when we do not need to.

boris-unckel replied:
@garydgregory The other cases before I understood the pattern behind "-1". In this case a FileInputStream is created, afterwards copyLarge calls copy calls copyLarge which then does the check.
I'll fix the former cases and wait with removal for cases like this on your comment.

garydgregory commented on src/main/java/org/apache/commons/io/FileUtils.java:
```diff
@@ -1200,6 +1210,7 @@ public static void deleteDirectory(final File directory) throws IOException {
      * @throws IOException          in case deletion is unsuccessful
      */
     private static void deleteDirectoryOnExit(final File directory) throws IOException {
+        Objects.requireNonNull(directory,"directory");
```
Fix formatting.

garydgregory commented on src/main/java/org/apache/commons/io/FilenameUtils.java:
```diff
@@ -1510,6 +1510,9 @@ public static boolean wildcardMatch(final String fileName, final String wildcard
      * @return true if the given name is a valid host name
      */
     private static boolean isValidHostName(final String name) {
+        if(Objects.isNull(name)) {
```
Fix formatting and no need to call an API meant for use in lambdas when `name == null` would do.

garydgregory commented on src/main/java/org/apache/commons/io/monitor/FileAlterationObserver.java:
```diff
@@ -200,12 +201,9 @@ public FileAlterationObserver(final File directory, final FileFilter fileFilter,
-            throw new IllegalArgumentException("Root directory is missing");
-        }
+        Objects.requireNonNull(rootEntry, "rootEntry");
+        Objects.requireNonNull(rootEntry.getFile(),"rootDirectory");
```
Fix formatting.

garydgregory commented:
> @garydgregory Please review and mark the cases you don't want to change for compatibility and/or design reasons. I'll change then all occurrences accordingly.

Hi @boris-unckel 

Sorry but I don't want to take the time to review each change in each of 31 files. If you do not want to revise the PR to minimize changes and duplication of null checks, that's fine as well. I might get to these kinds of changes later. Thank you for your contributions so far though :-)

boris-unckel commented:
@garydgregory 
- Fixed the formatting (is there a official eclipse configuration you can recommend?)
- Removed single parameter checks in case of direct usage, 
- Kept single parameter checks when prior (different) Object creation happens (see case above)
- Kept multiple parameter checks
- Removed double checks in constructors, even if objects are created
- Kept all checks in constructors, where fail will be in post constructor usage of methods

garydgregory commented:
@boris-unckel 
May you please rebase on master?
